pkg/alluxiocluster/controller.go

```diff
@@ -32,88 +33,92 @@ type AlluxioClusterReconciler struct {
 }
 
 type AlluxioClusterReconcileReqCtx struct {
-	*alluxiov1alpha1.AlluxioCluster
+	AlluxioClusterer
 	client.Client
 	context.Context
-	*alluxiov1alpha1.Dataset
+	datasetPkg.Datasetter
 	types.NamespacedName
 }
 
 func (r *AlluxioClusterReconciler) Reconcile(context context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger.Infof("Reconciling AlluxioCluster %s", req.NamespacedName.String())
-	ctx := AlluxioClusterReconcileReqCtx{
+	ctx := &AlluxioClusterReconcileReqCtx{
 		Client:         r.Client,
 		Context:        context,
 		NamespacedName: req.NamespacedName,
 	}
 
 	alluxioCluster := &alluxiov1alpha1.AlluxioCluster{}
-	ctx.AlluxioCluster = alluxioCluster
-	if err := r.Get(context, req.NamespacedName, alluxioCluster); err != nil {
-		if errors.IsNotFound(err) {
-			logger.Infof("Alluxio cluster %s not found. It is being deleted or already deleted.", req.NamespacedName.String())
-		} else {
-			logger.Errorf("Failed to get Alluxio cluster %s: %v", req.NamespacedName.String(), err)
-			return ctrl.Result{}, err
-		}
+	ctx.AlluxioClusterer = alluxioCluster
+	if err := GetAlluxioClusterFromK8sApiServer(r, req.NamespacedName, alluxioCluster); err != nil {
+		return ctrl.Result{}, err
 	}
+
 	dataset := &alluxiov1alpha1.Dataset{}
 	datasetNamespacedName := types.NamespacedName{
-		Name:      alluxioCluster.Spec.Dataset,
+		Name:      *alluxioCluster.Spec.Dataset,
```

Comment:
For this particular field "Dataset" it is required, and if it's not specified, the config file wouldn't pass the check and would fail before any code we wrote is executed. But I see your point here and let me see if there's anything else we haven't checked the dereference.